src/back-end/lib/db.ts

```diff
@@ -668,6 +669,33 @@ export async function hasFilePermission(connection: Connection, session: Session
   }
 }
 
+export async function hasAttachmentPermission(connection: Connection, session: Session, id: string): Promise<boolean> {
+  // If file is an attachment on a publicly viewable opportunity, allow
+  const results = await connection('cwuOpportunityAttachments as attachments')
+    .innerJoin('cwuOpportunityVersions as versions', 'versions.id', '=', 'attachments.opportunityVersion')
+    .innerJoin('cwuOpportunityStatuses as statuses', 'statuses.opportunity', '=' , 'versions.opportunity')
+    .whereIn('statuses.status', publicOpportunityStatuses as CWUOpportunityStatus[])
+    .andWhere({ 'attachments.file': id })
+    .select('attachments.*');
+
+  if (results.length > 0) {
+    return true;
+  }
+
+  // If file is an attachment on a proposal, and requesting user has access to the proposal, allow
+  if (session.user) {
+    const proposals = await connection('cwuProposalAttachments as attachments')
+      .innerJoin('cwuProposals as proposals', 'proposals.id', '=', 'attachments.proposal')
+      .where({ 'attachments.file': id })
+      .select<RawCWUProposal[]>('proposals.*');
+
+    if (proposals.length > 0) {
+      return proposals.some(async proposal => await hasReadPermissionCWUProposal(connection, session, proposal.opportunity, proposal.id));
```

Comment:
Does `.some` work with `async/await` ?

Reply:
If not, you'll need to use a `for...of` loop

Reply:
Smart use of the `permissions` function btw

Reply:
Yeah, good catch, it's not clear that will always behave.  I've switched it to a for...of for clarity.

Reply:
Generally, built-in iterative functions with callbacks are synchronous and don't support async/await. If you need to iterate over a list and run async code, you should use either map & Promise.all, or a `for...of` loop.